ethree-enclave/src/main/java/com/virgilsecurity/android/ethreeenclave/interaction/EThree.kt

```diff
@@ -0,0 +1,136 @@
+ *  * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ *  * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ *  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.virgilsecurity.android.ethreeenclave.interaction
+
+import android.content.Context
+import com.virgilsecurity.android.common.Const.NO_CONTEXT
+import com.virgilsecurity.android.common.callback.OnGetTokenCallback
+import com.virgilsecurity.android.common.interaction.EThreeCore
+import com.virgilsecurity.android.common.interaction.KeyManagerLocal
+import com.virgilsecurity.android.common.model.Result
+import com.virgilsecurity.sdk.androidutils.storage.AndroidKeyEntry
+import com.virgilsecurity.sdk.androidutils.storage.AndroidKeyStorage
+import com.virgilsecurity.sdk.jwt.Jwt
+import com.virgilsecurity.sdk.jwt.accessProviders.CachingJwtProvider
+import com.virgilsecurity.sdk.jwt.contract.AccessTokenProvider
+import com.virgilsecurity.sdk.storage.DefaultKeyStorage
+
+/**
+ * [EThree] class simplifies work with Virgil Services to easily implement End to End Encrypted
+ * communication.
+ */
+class EThree
+private constructor(
+        context: Context,
+        tokenProvider: AccessTokenProvider,
+        alias: String,
+        isAuthenticationRequired: Boolean,
+        keyValidityDuration: Int
+) : EThreeCore(tokenProvider) {
+
+    override val keyManagerLocal: KeyManagerLocal
+
+    init {
+        synchronized(this@EThree) {
+            val keyStorageAndroid = AndroidKeyStorage.Builder(alias)
+                    .isAuthenticationRequired(isAuthenticationRequired)
+                    .withKeyValidityDuration(keyValidityDuration)
+                    .onPath(context.filesDir.absolutePath)
+                    .build()
+
+            keyManagerLocal = KeyManagerLocalEnclave(keyStorageAndroid,
+                                                     tokenProvider.getToken(NO_CONTEXT).identity)
+
+            // Migration from old storage to new
+            val keyStorageDefault = DefaultKeyStorage(context.filesDir.absolutePath, KEYSTORE_NAME)
+            keyStorageDefault.names().forEach { name ->
+                val keyEntry = keyStorageDefault.load(name)
+
+                if (keyEntry.meta.isEmpty()) {
+                    val keyEntryAndroid = AndroidKeyEntry(name, keyEntry.value).apply {
+                        meta = mapOf("Migrated" to "true")
+                    }
+
+                    // If any error happens - restore state of storages.
+                    try {
+                        keyStorageAndroid.store(keyEntryAndroid)
+                    } catch (throwable: Throwable) {
+                        keyStorageAndroid.names().forEach { keyStorageAndroid.delete(it) }
```

Comment:
How user could know that his keys were not moved to encrypted keystorage?

Reply:
Good point